src/main/scala/pbdirect/PBScalarValueWriter.scala

```diff
@@ -54,85 +70,131 @@ trait LowPriorityPBScalarValueWriterImplicits {
+      override def wireType: Int                   = WIRETYPE_VARINT
+      override def isDefault(value: Long): Boolean = value == 0L
+      override def writeWithoutTag(value: Long, out: CodedOutputStream): Unit =
+        out.writeInt64NoTag(value)
     }
 
   implicit val floatWriter: PBScalarValueWriter[Float] =
-    instance(WIRETYPE_FIXED32) { (value, out) =>
-      out.writeFloatNoTag(value)
+    new PBScalarValueWriter[Float] {
+      override def wireType: Int                    = WIRETYPE_FIXED32
+      override def isDefault(value: Float): Boolean = value == 0.0F
+      override def writeWithoutTag(value: Float, out: CodedOutputStream): Unit =
+        out.writeFloatNoTag(value)
     }
 
   implicit val doubleWriter: PBScalarValueWriter[Double] =
-    instance(WIRETYPE_FIXED64) { (value, out) =>
-      out.writeDoubleNoTag(value)
+    new PBScalarValueWriter[Double] {
+      override def wireType: Int                     = WIRETYPE_FIXED64
+      override def isDefault(value: Double): Boolean = value == 0.0
+      override def writeWithoutTag(value: Double, out: CodedOutputStream): Unit =
+        out.writeDoubleNoTag(value)
     }
 
   implicit val stringWriter: PBScalarValueWriter[String] =
-    instance(WIRETYPE_LENGTH_DELIMITED) { (value, out) =>
-      out.writeStringNoTag(value)
+    new PBScalarValueWriter[String] {
+      override def wireType: Int                     = WIRETYPE_LENGTH_DELIMITED
+      override def isDefault(value: String): Boolean = value.isEmpty
+      override def writeWithoutTag(value: String, out: CodedOutputStream): Unit =
+        out.writeStringNoTag(value)
     }
 
   implicit val bytesWriter: PBScalarValueWriter[Array[Byte]] =
-    instance(WIRETYPE_LENGTH_DELIMITED) { (value, out) =>
-      out.writeByteArrayNoTag(value)
+    new PBScalarValueWriter[Array[Byte]] {
+      override def wireType: Int                          = WIRETYPE_LENGTH_DELIMITED
+      override def isDefault(value: Array[Byte]): Boolean = value.isEmpty
+      override def writeWithoutTag(value: Array[Byte], out: CodedOutputStream): Unit =
+        out.writeByteArrayNoTag(value)
     }
 
   implicit def keyValuePairWriter[K, V](
-      implicit keyFieldWriter: PBFieldWriter[K],
-      valueFieldWriter: PBFieldWriter[V]): PBScalarValueWriter[(K, V)] =
-    instance(WIRETYPE_LENGTH_DELIMITED) { (pair, out) =>
-      val buffer    = new ByteArrayOutputStream()
-      val bufferOut = CodedOutputStream.newInstance(buffer)
-      keyFieldWriter.writeTo(1, pair._1, bufferOut)
-      valueFieldWriter.writeTo(2, pair._2, bufferOut)
-      bufferOut.flush()
-      out.writeByteArrayNoTag(buffer.toByteArray)
+      implicit keyWriter: PBScalarValueWriter[K],
+      valueWriter: PBScalarValueWriter[V]): PBScalarValueWriter[(K, V)] =
+    new PBScalarValueWriter[(K, V)] {
+      override def wireType: Int                    = WIRETYPE_LENGTH_DELIMITED
+      override def isDefault(pair: (K, V)): Boolean = false
```

Comment:
Just out of curiosity, why the asymmetry compared to https://github.com/47deg/pbdirect/pull/27/files#diff-d48103461043e2a81b6050f08b04d8bbR118 ?

Reply:
I don't think the idea of a "default" key-value pair, so I arbitrarily set it to return false here. In the read code that you linked, I implemented it that way because I needed to return some (K, V).

Actually this code is never used so it doesn't really matter:
* when we write a map, we want to write all key-value pairs, even if any of the key-value pairs happens to be 'default'
* when we read a map, we simply read all the (zero or more) key-value pairs we find, so we have no need to fall back to a default value

But you're right to point that the asymmetry is confusing. I'll update the writer code to make it consistent with what happens in the reader.

I think you've hit upon a slight weirdness in my design. I'm treating things like embedded messages and key-value pairs as scalar values, which is not strictly correct. But I think this is close enough to reality for now. We can always refactor later.

Reply:
thank you for the answer :+1: